x/delayedack/keeper/fraud.go

```diff
@@ -0,0 +1,61 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	commontypes "github.com/dymensionxyz/dymension/v3/x/common/types"
+)
+
+func (k Keeper) HandleFraud(ctx sdk.Context, rollappID string) error {
+	// Get all the pending packets
+	// TODO (#631): Prefix store by rollappID for efficient querying
+	rollappPendingPackets := k.ListRollappPacketsByStatus(ctx, commontypes.Status_PENDING, 0)
+	if len(rollappPendingPackets) == 0 {
+		return nil
+	}
+	logger := ctx.Logger().With("module", "DelayedAckMiddleware")
+	logger.Debug("Reverting IBC rollapp packets", "rollappID", rollappID)
+	for _, rollappPacket := range rollappPendingPackets {
+		if rollappPacket.RollappId != rollappID {
+			continue
+		}
+
+		errString := "fraudulent packet"
+		packetId := channeltypes.NewPacketID(rollappPacket.Packet.GetDestPort(), rollappPacket.Packet.GetDestChannel(), rollappPacket.Packet.GetSequence())
+		logger.Debug("Reverting IBC rollapp packet", "rollappID", rollappID, "packetId", packetId, "type", rollappPacket.Type)
+
+		if rollappPacket.Type == commontypes.RollappPacket_ON_RECV {
+			err := k.writeFailedAck(ctx, rollappPacket, errString)
+			if err != nil {
+				logger.Error("failed to write failed ack", "rollappID", rollappID, "packetId", packetId, "error", errString)
+				//don't return here as it's nice to have
+			}
+		}
+
+		// Update status to reverted
+		rollappPacket.Error = errString
+		rollappPacket, err := k.UpdateRollappPacketWithStatus(ctx, rollappPacket, commontypes.Status_REVERTED)
+		if err != nil {
+			logger.Error("Error reverting IBC rollapp packet", "rollappID", rollappID, "packetId", packetId, "type", rollappPacket.Type, "error", err.Error())
+			return err
```

Comment:
here we'll revert fraud proof on first packet failure. possibly worth wrapping in applyFuncIfNoErr and continue to avoid one packet failing the reversion for all other. especially as the `updateRollappPacketWithStatus` triggers various hook so doesn't necessarily mean store error. wdyt

Reply:
thought about it quite long.
I just think that the alternative of keeping `Pending` packets of fraudulent rollapp may cause unexpected behaviour. maybe break invariant